4.gene_expression_signatures/scripts/0.latent_dimension_dataframe.py

```diff
@@ -0,0 +1,283 @@
+    gene_dict_df["qc_pass"], "dependency_column"
+].tolist()
+
+weight_data = weight_df.filter(gene_list_passed_qc, axis=1)
+weight_data.head()
+
+
+# In[4]:
+
+
+# Convert dataframes to tensors
+train_tensor = torch.tensor(train_df, dtype=torch.float32)
+test_tensor = torch.tensor(test_df, dtype=torch.float32)
+val_tensor = torch.tensor(val_df, dtype=torch.float32)
+
+
+# In[5]:
+
+
+output_dir = pathlib.Path("./results").resolve()
+output_dir.mkdir(parents=True, exist_ok=True)
+
+
+# In[6]:
+
+
+# Function to perform dimensionality reduction and extract z matrices for sklearn models
+def extract_z_matrix_sklearn(model, data, model_name, num_components, output_dir):
+    if model_name == "nmf":
+        min_value = data.min().min()
+        if min_value < 0:
+            data = data - min_value
+    # Fit the model and transform the data
+    z_matrix = model.fit_transform(data)
+    
+    # Create a DataFrame and save it
+    z_matrix_df = pd.DataFrame(z_matrix, columns=[f'{model_name}_{i}' for i in range(num_components)])
+    z_matrix_df.insert(0, 'ModelID', train_data.index) 
+    
+    return z_matrix_df
+
+
+# In[7]:
+
+
+def extract_weights(model, model_name, H=None):
+    if model_name in ["pca", "ica", "nmf"]:
+        # Transpose PCA components and format columns
+        weights_df = pd.DataFrame(model.components_, columns=dependency_df.drop(columns=["ModelID"]).columns.tolist()).transpose()
+        weights_df.columns = [f"{x}" for x in range(0, weights_df.shape[1])]
+        
+    # Reset index to rename 'index' to 'genes'
+    weights_df = weights_df.reset_index().rename(columns={"index": "genes"})
+    
+    return weights_df
+
+
+def perform_gsea(weights_df, model_name, num_components):
+    library = blitz.enrichr.get_library("Reactome_2022")
+    seed = random.random()
+    gsea_results = []
+    # Define cutoff values
+    lfc_cutoff = 0.584
+    fdr_cutoff = 0.25
+    # Perform GSEA for each component column in weights_df
+    for col in weights_df.columns[1:]:  # Skip 'genes' column
+        gene_signature = weights_df[['genes', col]]
+        
+        if gene_signature.shape[0] > 0:
+            try:
+                # Perform GSEA using the gene signature (weights)
+                gsea_result = blitz.gsea(gene_signature, library, seed=seed)
+                gsea_result = gsea_result.reset_index()
+                for _, pathway_result in gsea_result.iterrows():
+                    result_row = {
+                        "z": int(col),
+                        "full_model_z": num_components,
+                        "model": str(model_name),
+                        "reactome pathway": str(pathway_result['Term']),
+                        "gsea es score": pathway_result['es'],
+                        "nes score": pathway_result['nes'],
+                        "p value": pathway_result['pval'],
+                        "shuffled": False
+                    }
+                    gsea_results.append(result_row)
```

Comment:
Not sure if you need to do all of this? Is `gsea_result` already a pandas dataframe? Is this loop renaming columns and adding new ones?

Can you do something like this then?

```python
gsea_result = (
    gsea_result.rename(
        columns={
            "model": "model_rename"
        }
    )
    .assign(shuffled=False)
)
```